Lib/code.py

```diff
@@ -150,10 +150,21 @@ def showtraceback(self, **kwargs):
             else:
                 # If someone has set sys.excepthook, we let that take precedence
                 # over self.write
-                sys.excepthook(ei[0], ei[1], last_tb)
+                self._call_excepthook(ei[0], ei[1], last_tb)
         finally:
             last_tb = ei = None
 
+    def _call_excepthook(self, typ, value, tb):
+        try:
+            sys.excepthook(typ, value, tb)
+        except Exception as e:
+            e.__context__ = None
+            print('Error in sys.excepthook:', file=sys.stderr)
+            sys.__excepthook__(type(e), e, e.__traceback__.tb_next)
+            print(file=sys.stderr)
+            print('Original exception was:', file=sys.stderr)
```

Comment:
What will happen if this code raise an exception?

Reply:
I don't know whether it makes sense to wrap this code (and the `traceback.format_exception` calls) into more `try...excepts`. I think we have three options:

- add `try...excepts` carefully at various points in `code.py`, with the risk of missing cases.
- add an `except BaseException` in the repl loop in `InteractiveConsole.interact`. that feels pretty invasive to me and I'm not sure it could plausibly be backported as a bug fix.
- add an `except BaseException` in the pyrepl loop in `run_multiline_interactive_console`. this would solve the problem of pyrepl crashing if you do weird things in a more targeted way (but not fix any of these problems in `code`).

I would be leaning towards option 3) (because the classic repl is incredibly robust against doing weird things, including closing `sys.stderr` and stuff like that, so pyrepl should be too). but I definitely think that it is beyond the scope of this specific issue/PR.

Reply:
I've opened #122468 to discuss the problem of crashes in pyrepl in general. imo the problem of `sys.excepthook` in `code` is a bit independent (the bug long ago came from a real-world issue that people had on pypy).